src/main/java/org/candlepin/subscriptions/resource/api/v2/SubscriptionTableController.java

```diff
@@ -0,0 +1,322 @@
+import org.candlepin.subscriptions.db.SubscriptionCapacityViewRepository;
+import org.candlepin.subscriptions.db.model.BillingProvider;
+import org.candlepin.subscriptions.db.model.HardwareMeasurementType;
+import org.candlepin.subscriptions.db.model.ServiceLevel;
+import org.candlepin.subscriptions.db.model.SubscriptionCapacityView;
+import org.candlepin.subscriptions.db.model.SubscriptionCapacityViewMetric;
+import org.candlepin.subscriptions.db.model.Usage;
+import org.candlepin.subscriptions.resource.ResourceUtils;
+import org.candlepin.subscriptions.utilization.api.model.BillingProviderType;
+import org.candlepin.subscriptions.utilization.api.model.ReportCategory;
+import org.candlepin.subscriptions.utilization.api.model.ServiceLevelType;
+import org.candlepin.subscriptions.utilization.api.model.SkuCapacitySubscription;
+import org.candlepin.subscriptions.utilization.api.model.SortDirection;
+import org.candlepin.subscriptions.utilization.api.model.SubscriptionEventType;
+import org.candlepin.subscriptions.utilization.api.model.SubscriptionType;
+import org.candlepin.subscriptions.utilization.api.model.UsageType;
+import org.candlepin.subscriptions.utilization.api.v2.model.SkuCapacity;
+import org.candlepin.subscriptions.utilization.api.v2.model.SkuCapacityReport;
+import org.candlepin.subscriptions.utilization.api.v2.model.SkuCapacityReportMeta;
+import org.candlepin.subscriptions.utilization.api.v2.model.SkuCapacityReportSort;
+import org.springframework.data.domain.Pageable;
+import org.springframework.stereotype.Service;
+
+@Service(value = "v2SubscriptionTableController")
+@Slf4j
+@AllArgsConstructor
+public class SubscriptionTableController {
+
+  private final SubscriptionCapacityViewRepository repository;
+  private final ApplicationClock clock;
+
+  @Transactional
+  public SkuCapacityReport capacityReportBySku( // NOSONAR
+      ProductId productId,
+      @Min(0) Integer offset,
+      @Min(1) Integer limit,
+      ReportCategory category,
+      ServiceLevelType serviceLevel,
+      UsageType usage,
+      BillingProviderType billingProviderType,
+      String billingAccountId,
+      String metricId,
+      SkuCapacityReportSort sort,
+      SortDirection dir) {
+    var subscriptionSpec =
+        SubscriptionCapacityViewRepository.buildSearchSpecification(
+            getOrgId(),
+            productId,
+            category,
+            serviceLevel,
+            usage,
+            billingProviderType,
+            billingAccountId,
+            metricId);
+    MetricsMeta metrics = new MetricsMeta(metricId, productId);
+    Map<String, SkuCapacity> reportItemsBySku = new HashMap<>();
+    repository
+        .streamBy(subscriptionSpec)
+        .forEach(
+            subscription -> {
+              String inventoryKey = subscription.getSku();
+              if (productId.isPrometheusEnabled()) {
+                inventoryKey += ":" + subscription.getBillingProvider();
+              }
```

Comment:
Rather than encoding into a string like this, I'd rather see a small data class having sku and billing provider as fields.

Reply:
updated